test/gtest/test_basic.cc

```diff
@@ -5,10 +5,285 @@
+}
+
+TEST_F(test_basic, xpmem_make_unaligned_failure) {
+  xpmem_segid_t segid =
+      xpmem_make((void*)12, 1 << 16, XPMEM_PERMIT_MODE, (void*)0600);
+  ASSERT_EQ(-1, segid);
+}
+
+TEST_F(test_basic, xpmem_make_perm_failure) {
+  xpmem_segid_t segid = xpmem_make(NULL, 1 << 16, 0, (void*)0600);
+  ASSERT_EQ(-1, segid);
+}
+
+TEST_F(test_basic, xpmem_make_overflow_failure) {
+  xpmem_segid_t segid =
+      xpmem_make((void*)-2, 1 << 16, XPMEM_PERMIT_MODE, (void*)0600);
+  ASSERT_EQ(-1, segid);
+}
+
+TEST_F(test_basic, xpmem_remove_error) { ASSERT_EQ(-1, xpmem_remove(-1)); }
+
+TEST_F(test_basic, xpmem_remove_unknown) { ASSERT_EQ(-1, xpmem_remove(12345)); }
+
+TEST_F(test_basic, xpmem_remove_twice) {
+  xpmem_segid_t segid =
+      xpmem_make(NULL, 10 * page_size(), XPMEM_PERMIT_MODE, (void*)0600);
+  ASSERT_NE(-1, segid);
+  ASSERT_EQ(0, xpmem_remove(segid));
+  ASSERT_EQ(-1, xpmem_remove(segid));
+}
+
+TEST_F(test_basic, xpmem_get_bad_segid) {
+  xpmem_apid_t apid = xpmem_get(-1, XPMEM_RDWR, XPMEM_PERMIT_MODE, (void*)0600);
+  ASSERT_EQ(-1, apid);
+}
+
+TEST_F(test_basic, xpmem_get_bad_segid_positive) {
+  xpmem_apid_t apid =
+      xpmem_get(12345, XPMEM_RDWR, XPMEM_PERMIT_MODE, (void*)0600);
+  ASSERT_EQ(-1, apid);
+}
+
+TEST_F(test_basic, xpmem_get_success) {
+  xpmem_segid_t segid =
+      xpmem_make(NULL, 10 * page_size(), XPMEM_PERMIT_MODE, (void*)0600);
+  ASSERT_NE(-1, segid);
+
+  xpmem_apid_t apid =
+      xpmem_get(segid, XPMEM_RDWR, XPMEM_PERMIT_MODE, (void*)0600);
+  ASSERT_NE(-1, apid);
+  ASSERT_EQ(0, xpmem_release(apid));
+  ASSERT_EQ(0, xpmem_remove(segid));
+}
+
+TEST_F(test_basic, xpmem_get_bad_permit_mode) {
+  xpmem_segid_t segid =
+      xpmem_make(NULL, 10 * page_size(), XPMEM_PERMIT_MODE, (void*)0200);
+  ASSERT_NE(-1, segid);
+
+  xpmem_apid_t apid = xpmem_get(segid, 9999, XPMEM_PERMIT_MODE, (void*)0);
+  EXPECT_EQ(-1, apid);
+
+  ASSERT_EQ(0, xpmem_remove(segid));
+}
+
+// Missing: Group permission checks
+// Missing: Seems that xpmem_attach() anyways only allows XPMEM_RDWR
+TEST_F(test_basic, xpmem_get_not_allowed_permit) {
+  auto create_xpmem = [this](uintptr_t make_perms, uintptr_t get_perms) {
+    xpmem_segid_t segid = xpmem_make(NULL, 10 * page_size(), XPMEM_PERMIT_MODE,
+                                     (void*)make_perms);
+    xpmem_apid_t apid = -1;
+    if (segid != -1) {
+      apid = xpmem_get(segid, get_perms, XPMEM_PERMIT_MODE, NULL);
+    }
+    return std::make_pair(segid, apid);
+  };
+
+  auto mem = create_xpmem(0600, XPMEM_RDONLY);
+  ASSERT_NE(-1, mem.first);
+  EXPECT_NE(-1, mem.second);
+  ASSERT_EQ(0, xpmem_remove(mem.first));
+
+  std::vector<std::pair<uintptr_t, uintptr_t>> todo = {
+    {0200, XPMEM_RDWR},
+    {0000, XPMEM_RDWR},
+    {0000, XPMEM_RDONLY},
+  };
+  for (auto pair : todo) {
+    mem = create_xpmem(pair.first, pair.second);
+    ASSERT_NE(-1, mem.first);
+    EXPECT_EQ(-1, mem.second);
+    ASSERT_EQ(0, xpmem_remove(mem.first));
+  }
+}
+
+struct xpmem_context {
```

Comment:
maybe better use class for consistency with other tests?

Reply:
fixed